transformations/transformations/pool_allocator.py

```diff
@@ -659,25 +695,49 @@ def create_pool_allocator(self, routine, stack_size):
+            if call.name in targets or call.routine.name.lower() in ignore: #['surfpp', 'surfexcdriver', 'surfseb']:
                # If call is declared via an explicit interface, the ProcedureSymbol corresponding to the call is the
                # interface block rather than the Subroutine itself. This means we have to update the interface block
                # accordingly
                 if call.name in [s for i in FindNodes(Interface).visit(routine.spec) for s in i.symbols]:
                     _ = self._get_stack_arg(call.routine)
-
-                if call.routine != BasicType.DEFERRED and self.stack_argument_name in call.routine.arguments:
-                    arg_idx = call.routine.arguments.index(self.stack_argument_name)
-                    arguments = call.arguments
-                    call_map[call] = call.clone(arguments=arguments[:arg_idx] + (stack_var,) + arguments[arg_idx:])
-
+                if call.routine != BasicType.DEFERRED and self.stack_local_var_name not in call.arguments:
+                    if self.stack_argument_name in call.routine.arguments:
+                        arg_idx = call.routine.arguments.index(self.stack_argument_name)
+                        arguments = call.arguments
+                        kwarguments = call.kwarguments
+                        if arg_idx <= len(arguments):
+                            arguments = arguments[:arg_idx] + (stack_var,) + arguments[arg_idx:]
+                        else:
+                            arg_idx -= len(arguments)
+                            kwarguments = kwarguments[:arg_idx] + ((self.stack_argument_name, stack_var),) + \
+                                          kwarguments[arg_idx:]
+                        call_map[call] = call.clone(arguments=arguments, kwarguments=kwarguments)
+                    else:
+                        arguments = call.arguments
+                        kwarguments = call.kwarguments
+                        arg_pos = [call.routine.arguments.index(arg) for arg in call.routine.arguments
+                                   if arg.type.optional]
+                        if arg_pos:
+                            if arg_pos[0] < len(arguments):
+                                arguments = arguments[:arg_pos[0]] + (stack_var,) + arguments[arg_pos[0]:]
+                            else:
+                                arg_idx = arg_pos[0] - len(arguments)
+                                kwarguments = kwarguments[:arg_idx] + ((self.stack_argument_name, stack_var),) + \
+                                              kwarguments[arg_idx:]
+                        else:
+                            if kwarguments:
+                                kwarguments += ((self.stack_argument_name, stack_var),)
+                            else:
+                                arguments += (stack_var,)
+                        call_map[call] = call.clone(arguments=as_tuple(arguments), kwarguments=as_tuple(kwarguments))
```

Comment:
What's going on here? I suspect this has to do with picking the right injection point when there are keyword arguments present?
That index-fuffing is a bit hard to follow, could this be made more elegant or maybe even outlined into a utility routine (because I think we discussed this interface update problem in a wider context before)?